### DISCUSSION TITLE
Find largest interesting area without size requirements

### DISCUSSION THREAD

**Discussion by L3tum** [libvips/libvips#3376] [Q&A] [answered]
Hey, I'm currently, in essence, trying to identify the largest area that a `vips_smartcrop` operation would cut out. 
Since smartcrop requires pixel dimensions to be passed in, there is no way (that I found or know of) to get as large an area as possible.

A specific example would be a landscape image that has large areas of just background colour on either side, which could be cut off when resizing the image to a portrait image (and then filling the result in at the top and bottom to make it the required size). 

So the pipeline would be:
- base image
- find out boring areas or largest interesting area
- cut the base image to these specifications
- fill the image up to the right aspect ratio at the top and bottom
- rescale the image to the proper pixel dimensions

Step 2 is essentially the only one I have no idea how to do. It's not necessarily a uniform background colour, just a very low entropy.

**Comment by jcupitt** [comment]
Hi @L3tum,

Could you suggest a sample image and the result you'd like? It helps to make discussion more concrete.

**Reply by L3tum** [reply]
Sure, let's say you have an image like this:

![Netflix_Games](https://user-images.githubusercontent.com/9307432/224075128-f2a9e75c-38a4-44e8-b3b8-1767b4f71d20.png)

I want to know that the only meaningful content, i.e. the area with high entropy, is this:
![Netflix_Games](https://user-images.githubusercontent.com/9307432/224075476-9fa01720-cc45-4369-b471-d69ca58e7a83.png)

So that I can cut it like this:
![Netflix_Games](https://user-images.githubusercontent.com/9307432/224075817-4fc8f383-d852-4453-81c5-b9d1f8fe202f.png)

And then resize + fill it like this:
![Netflix_Games](https://user-images.githubusercontent.com/9307432/224076381-98f01e27-95ea-482e-8979-3471ac1cbf7d.png)


Most of the steps are somewhat easy to do. The only issue I have is finding the area with high entropy. Smartcrop only accept a given size, and I essentially want it to tell me instead how much it thinks it should keep of the image. If I tell it that I want the image in 100x200 then it may cut off something that I still want in the image, because the meaningful content is 103x200. I'd rather fill some of the image with some background colour than cut off to much.

(FYI I'm not affiliated with Netflix, just the first image that somewhat fit the theme concerning the overall makeup of the image in regards to background colour and such)

**Comment by jcupitt** [comment, marked as answer]
I think I'd use `find_trim` instead for this specific case. Do a bit of filtering to remove smooth gradients, then `find_trim` ought to do a good job, eg.:

![image](https://user-images.githubusercontent.com/580843/224091226-247c82e0-daa6-4b04-b143-3141d7cbedaf.png)

**Reply by L3tum** [reply]
Thanks, that looks pretty good! Is there an equivalent for the custom blur in php-vips? I've looked at the source in nip2 and built the following based on it, but `find_trim` returns the original image dimensions. The $processed image actually looks the same as the one in your example. I've also tried leaving the background out or playing around with the threshold but it doesn't seem to change anything. 

````php
$image = Image::newFromFile($source);
$fac = 1.0;
$radius = 45.0;
$gaussMask = Image::gaussmat($radius / 3.0, 0.2, ['separable' => true]);
$blur = $image->convsep($gaussMask);
$processed = $blur->multiply($fac)->add(($image->multiply(1 - $fac)));
$processed->writeToFile($targetIntermediate);

$processed = $image->subtract($processed);

$processed->writeToFile($targetIntermediate2);

$cropArea = $processed->find_trim(['background' => [0,0,0]]);
````

**Reply by jcupitt** [reply]
You can use `gaussblur()` to blur. eg.:

```php
#!/usr/bin/php
<?php

require __DIR__ . '/vendor/autoload.php';
use Jcupitt\Vips;

$image = Vips\Image::newFromFile($argv[1]);
$crop = $image->subtract($image->gaussblur(40))->find_trim(["threshold" => 30, "background" => 0]);

echo "crop = ";
print_r($crop);
echo "\n";
```

I see:

```
john@banana ~/try $ ./find_crop.php ~/pics/netflix.png 
crop = Array
(
    [left] => 371
    [top] => 123
    [width] => 216
    [height] => 297
)
```

**Reply by L3tum** [reply]
That did the trick! Thank you! You're a big help, I don't think I could've figured it out.

````php
 $image = Image::newFromFile($source);
$processed = $image->subtract($image->gaussblur(15));
$cropArea = $processed->find_trim(['threshold' => 100, 'background' => 0]);
$image = $image->crop($cropArea['left'], $cropArea['top'], $cropArea['width'], $cropArea['height']);
$image->xoffset = 0;
$image->yoffset = 0;
$image->writeToFile($target);
````

With this image I found:

![source.png](https://user-images.githubusercontent.com/9307432/224319053-0cd662d9-09ea-473c-9fa3-0a6dd5fd67ff.png)

Results in this crop:
![cropped.png](https://user-images.githubusercontent.com/9307432/224319137-8c88a1da-930c-4709-85eb-792b822510ec.png)


The hair is a bit cut off but that's just some fine-tuning to be done.
